### ISSUE TITLE
Bundles, descriptions and issues

### ISSUE THREAD

**Issue by polarkernel** [prochitecture/blosm#80]
The first task in the search for objects of the class `Bundle` is the search for equidistant path segments. For a better understanding of the concept and its problems, I will first explain how the algorithm works.

First, all way-sections that exceed a minimal length and are not too curved are entered into a static spatial index. This index uses bounding boxes around the sections to find neighbor sections. Way-sections, whose bounding boxes overlap, are candidates as neighbors. For every section (template) in this index, its neighbors are searched and tested, whether they are equidistant to the template section. This is done as follows:

<img src="https://github.com/prochitecture/blosm/assets/48180714/993ead89-2b59-4808-9c55-36d550119747" width="400">

The black line is the centerline of the template section. A buffering polygon is constructed around this line, spaced according to the category of the section (red dotted polygon). Assume that the cyan line is the centerline of the neighbor candidate. A clipping algorithm first determines the length `inLineLength` of the candidate within the buffer polygon. In this example, this is the length between `p1` and `p2`. This length must at least be 10 % of the total length of the candidate. 

However, a short candidate line could be perpendicular to the template, but almost completely within the buffer polygon. So we need an additional condition: The distances `d1` and `d2` between the entry and exit points `p1` and `p2` and the template are computed. Then, some kind of slope relative to the template is determined from the difference between these distances and the length `inLineLength` of the candidate within the buffer polygon to as `slope = abs(d1-d2)/inLineLength`. Demanding a small value for `slope` rules such cases out.

This algorithm has the advantage, that overlapping way-sections continue the equidistant parts, as long as only one side is interrupted. For instance these three way-sections (black lines with red dotted surrounding buffer polygon) would be considered to belong to the same group (`Bundle`):

<img src="https://github.com/prochitecture/blosm/assets/48180714/10aae07a-f20e-4f5b-a60c-7de4ca1d53ed" width="600">

However, there is a disadvantage, when the intersections are almost perpendicular, so that this effect does not work anymore. In the image below on the left (middle part in _rotterdam_01.osm_) this effect is clearly visible. Overlapping sections result in parallel ways through the intersection (all the same color). In the image on the right (intersection between Karl-Marx-Allee and Straße der Pariser Kommune in _berlin_karl_marx_allee.osm_), the ways are almost perpendicular, and the groups always contain only the short parallel parts (same color is same group).

<img src="https://github.com/prochitecture/blosm/assets/48180714/13681758-2251-4f16-b804-db8daa119f30" width="300"> 
<img src="https://github.com/prochitecture/blosm/assets/48180714/548c0181-da01-4a55-83bb-0f8d4ab4b707" width="295">

**Comment by vvoovv** [comment]
> However, a short candidate line could be perpendicular to the template, but almost completely within the buffer polygon. 

I don't understand this case. If the short candidate is perpendicular to the template and completely or always completely within the buffer polygon, then there is no intersection or only one intersection of the short candidate with the buffer polygon. So both `d1` and `d2` can not be calculated or only either of `d1` or `d2` can be calculated.

**Comment by vvoovv** [comment]
> A buffering polygon is constructed around this line.

On a separate note. To construct the buffering polygon, the original curve must be offsetted. Do you check for self-intersections when offsetting the curve? Do you process the self-intersections? I need this feature to render roadways and sidewalks correctly.

**Comment by polarkernel** [comment]
> I don't understand this case.

The case with the intersection points `p1` and `p2`, which I have drawn above, is only a general illustration. Often, the equidistant candidates are completely inside the buffer polygon, like `candidateA` in the figure below. Then, the endpoints of the candidate line are used as the entry and exit points of the "_inside_" part. Because the difference between the distances of these points to the template line is very small compared to the inside length of the candidate, the value of `slope` gets small.

In this example, you can also see why I called this value _slope_. If we straightened the template line to a straight horizontal line and drew a straight line using the distances `d1` and `d2` for the points `p1` and `p2` of the candidate line, the value of `slope` would be exactly the slope of the candidate line.

For the other case with `candidateB`, the difference of the distances from the entry and exit points on the candidate line to the template line is large compared to the inside length of the candidate, the value of `slope` becomes large. Such a line can't be accepted as parallel (equidistant).

<img src="https://github.com/prochitecture/blosm/assets/48180714/edd5ec4a-30be-443a-9e54-911897780073" width="400">

**Comment by polarkernel** [comment]
> Do you check for self-intersections when offsetting the curve? Do you process the self-intersections?

No, not completely. Self-intersections are only processed between adjacent line segments. For the equidistant line detection application, this is not important because lines that are too curved are already excluded from the static spatial index. In addition, I successfully replaced corners with a `Corner` class, as discussed at the end of https://github.com/prochitecture/blosm/issues/72#issuecomment-1874250570.

<img src="https://github.com/prochitecture/blosm/assets/48180714/68d2138a-c017-499a-889e-11826305a008" width="400">

Some time ago, I made an attempt to use a different algorithm that gave better results, as far as I can remember. I will try to find it and let you know if I do.

**Comment by vvoovv** [comment]
I actually meant this kind of self-intersections:

![image](https://github.com/prochitecture/blosm/assets/613295/6ca0f92e-91e5-41a3-95ba-9297049245f1)

**Comment by polarkernel** [comment]
> I actually meant this kind of self-intersections:

I think these are computed correctly, as expected for an offset line:

 <img src="https://github.com/prochitecture/blosm/assets/48180714/510c8515-7a04-4f03-a4e5-2c8127667bad" width="400">

**Comment by vvoovv** [comment]
> I think these are computed correctly, as expected for an offset line

Could you please paste the code to generate the image above. I'd like to know the API calls to generate an offset for a curve with processing of self-intersections.

**Comment by polarkernel** [comment]
The code was:
```Python
from mathutils import Vector
import matplotlib.pyplot as plt
from lib.CompGeom.PolyLine import PolyLine

line = [
    Vector((1.8,3)),
    Vector((1,1.8)),
    Vector((0.2,1)),
    Vector((0,0)),
    Vector((0.2,-1)),
    Vector((0.5,-2)),
    Vector((0.8,-3))
]

polyline = PolyLine(line)
polyline.plot('k',2)

for d in range(20):
    dist = 0.2*(d+1)
    offsetPoly = polyline.buffer(dist,dist)

    x = [n[0] for n in offsetPoly] + [offsetPoly[0][0]]
    y = [n[1] for n in offsetPoly] + [offsetPoly[0][1]]
    plt.plot(x[:-1],y[:-1],'r',linewidth=0.5)
    plt.plot(x[:-1],y[:-1],'b.',markersize=2)

plt.gca().axis('equal')
plt.show()
```

**Comment by vvoovv** [comment]
![image](https://github.com/prochitecture/blosm/assets/613295/fe6436fc-1fce-42b3-b03e-d9d3a0438e95)

I noticed that extra points are created for convex corners. That rounds the resulting corners. Is it possible to turn the rounding mode off?

**Comment by polarkernel** [comment]
> I noticed that extra points are created for convex corners. That rounds the resulting corners. Is it possible to turn the rounding mode off?

Well, for now just a patch: In module _lib/CompGeom/OffsetGenerator.py_, replace [line 91](https://github.com/prochitecture/blosm/blob/2a60848b75de69912989c2883d7fc2602d4a45bb/lib/CompGeom/OffsetGenerator.py#L91)
```Python
fillet = self.constructFillet(p2,seg[1],p3,abs(self.distance))
```
by 
```Python
fillet = None #self.constructFillet(p2,seg[1],p3,abs(self.distance))
```

Should you like to use this feature permanently, we will have to introduce some function argument.

**Comment by polarkernel** [comment]
> Some time ago, I made an attempt to use a different algorithm that gave better results, as far as I can remember. I will try to find it and let you know if I do.

I found this old code. It implements a totally different algorithm. It has a function for lines and a function for polygons. As far as I remember, there were some minor floating-point accuracy issues (see red arrow), but it actually works on a much larger parameter range than the one that is in the addon. It is also capable of creating holes, as shown in the examples below, which is not the case with the current one. I am convinced that I would be able to finish it and make it robust and reliable if needed.

<img src="https://github.com/prochitecture/blosm/assets/48180714/808d8874-df93-4c36-9f67-715e9e110ede" width="400"> <img src="https://github.com/prochitecture/blosm/assets/48180714/3332a17a-8392-4862-a89b-e23abbcf183d" width="400">

**Comment by vvoovv** [comment]
> However, there is a disadvantage, when the intersections are almost perpendicular, so that this effect does not work anymore.

![image](https://github.com/prochitecture/blosm/assets/613295/07786a5b-ffda-400a-ba78-cd9442ee87ab)

Did I understand the problem correctly that it's not possible to join the adjacent bundles for perpendicular way-sections since the buffering polygons do not intersect in that case?

Adjacent way-sections of the bundle can be checked if they are collinear with the way-sections of the bundle and that they also form a bundle.

**Comment by polarkernel** [comment]
> Did I understand the problem correctly that it's not possible to join the adjacent bundles for perpendicular way-sections since the buffering polygons do not intersect in that case?

I should have written _difference_ instead of _disadvantage_. For now, this is just an observation that may or may not play a role later. I will try to build a more complete set of observations within this thread. For the moment it does not make sense to build a theory, an intersection may also look like this (on the left of _osm_extracts/streets/milano_01.osm_):

<img src="https://github.com/prochitecture/blosm/assets/48180714/f06691a8-c957-4b44-8270-510f4957e4ab" width="200">

**Comment by polarkernel** [comment]
I made a first attempt to find clustered intersections. It is quite simple and works as follows:

- Collect all the endpoints of sections that have been identified as equidistant into a set so that each vertex appears only once.
- Apply the[ DBSCAN](https://en.wikipedia.org/wiki/DBSCAN) algorithm to this set of endpoints. This algorithm groups closely-spaced points (points with neighbors within a given maximum distance) into point clusters, which can then be used as nuclei for clustered intersections.

The result looks already quite promising:

<img src="https://github.com/prochitecture/blosm/assets/48180714/62cae529-c858-43d3-a62b-390bdaed4a1e" width="450">

Looking at the images of the intersection between _Karl-Marx-Allee_ and _Straße der Pariser Kommune_ in _berlin_karl_marx_allee.osm_ (left) and of the intersection in the middle part of _rotterdam_01.osm_ (right), you can notice a problem, that will have to be discussed (the dotted polygon is the convex hull of all point detected as one group):

<img src="https://github.com/prochitecture/blosm/assets/48180714/7adaf7b3-50d8-49ef-83bb-f978f1c9547f" width="300"> <img src="https://github.com/prochitecture/blosm/assets/48180714/3af6621d-f0fd-4e0d-9758-681dd141eaab" width="300">

In both cases, all intersections belonging to this cluster have been detected, including those between pedestrian and bicycle ways. The question now is, where do the bundles (`Bundle` class) and the intersections (`Intersection` class) start? What about crosswalks? Do they belong to the bundles, and what about their intersections? 

Filtering out all these details is a major task that now awaits us. Not all crossings are as classically shaped as these two examples above:

<img src="https://github.com/prochitecture/blosm/assets/48180714/9e187135-705e-4cca-a2b4-7568b41d7fc0" width="300">

**Comment by vvoovv** [comment]
> Not all crossings are as classically shaped

What if intersections with footways and cycleways are ignored? Will the intersections have a more classical shape?

**Comment by vvoovv** [comment]
> What about crosswalks? Do they belong to the bundles, and what about their intersections?

Originally I thought that a crosswalk is a part of a `Roadway` (see also https://github.com/prochitecture/blosm/issues/72#issuecomment-1846141098). Filets on intersections make the things more complicated. The question can be also raised if a filet should be a part of an intersection.

The crosswalks are located on the filets on the [intersection](https://www.openstreetmap.org/#map=19/52.51512/13.43184) below:

![image](https://github.com/prochitecture/blosm/assets/613295/decebd5e-f85c-4ddc-ba1d-190225466df2)

I'll think what to do with the filets in that context.

**Comment by polarkernel** [comment]
> What if intersections with footways and cycleways are ignored? Will the intersections have a more classical shape?

Yes. It should also be simple to detect them. I have more concerns about their assignment, as you also noted:

> The question can be also raised if a filet should be a part of an intersection.

Locating fillets on bundles can be ambiguous. For example, the one at the top right might belong to the upward or to the right-hand bundle.

**Comment by polarkernel** [comment]
> What if intersections with footways and cycleways are ignored? Will the intersections have a more classical shape?

I committed a version in which intersections between categories of minor roads such as "pedestrian", "track", "footpath", "path", "cycleway" or "bridleway" are distinguished from intersections with major roads. The areas of the two are drawn as convex hull in different saturations, and the intersections of the former are drawn as rectangles and the latter as circles:

<img src="https://github.com/prochitecture/blosm/assets/48180714/73db734e-f6cb-461d-9474-6fcba387a692" width="300"> 
<img src="https://github.com/prochitecture/blosm/assets/48180714/f06a6777-3658-4589-8c8d-f44e6a7ec293" width="300">

**Comment by vvoovv** [comment]
I also suggest drawing a convex hull for "roadway" intersections (primary, secondary, tertiary, residential, etc).

**Comment by polarkernel** [comment]
> I also suggest drawing a convex hull for "roadway" intersections (primary, secondary, tertiary, residential, etc).

This is already done. The major road intersections have a round dot and their convex hull is darker (or more saturated, large `alpha`) than the minor road intersections, which have a smaller square dot and a brighter (less saturated, small `alpha`) convex hull. Major road intersections are defined as those, that have at least one way with a category, that does not belong to the categories of minor roads (`pedestrian`, `track`, `footpath`, `path`, `cycleway`, `bridleway`).

I made this version for visualization purposes only, so we can discuss issues using real-world examples.

**Comment by vvoovv** [comment]
I meant crossing nodes where ALL ways are not minor roads (`pedestrian`, `track`, `footpath`, `path`, `cycleway`, `bridleway`).

**Comment by polarkernel** [comment]
> I meant crossing nodes where ALL ways are not minor roads (`pedestrian`, `track`, `footpath`, `path`, `cycleway`, `bridleway`).

I see. I made now a version, that produces an additional category of intersections, which have only major roads. These are plotted as circle dots and their convex hull color is fully saturated with `alpha=1`.

<img src="https://github.com/prochitecture/blosm/assets/48180714/2b484b8e-98f0-4e87-aa21-a8360b0981c5" width="300">

For many clusters, there are only two such major intersections, so it is not possible to draw a convex hull. These are then connected by a thick line:

<img src="https://github.com/prochitecture/blosm/assets/48180714/13d17e21-f60b-4ac1-956e-bd13a1e70fac" width="300">

The version with this feature is committed.

**Comment by vvoovv** [comment]
> I made now a version, that produces an additional category of intersections, which have only major roads. 

To be more precise, a crossing node can be shared by minor ways, but it must be shared by at least 3 major ways.

**Comment by polarkernel** [comment]
> To be more precise, a crossing node can be shared by minor ways, but it must be shared by at least 3 major ways.

OK, committed version adapted.

**Comment by vvoovv** [comment]
> To be more precise, a crossing node can be shared by minor ways, but it must be shared by at least 3 major ways.

Sorry. That sentence was not formulated precisely again. I meant _at least 3 major way-segments_ instead of _at least 3 major ways_.

**Comment by polarkernel** [comment]
> Sorry. That sentence was not formulated precisely again. I meant at least 3 major way-segments instead of at least 3 major ways.

To be precise, here are the rules of the current implementation. For each node that is an intersection, the sum  `wayCount` of all incoming and outgoing ways is determined. The number of ways belonging to the categories `pedestrian`, `track`, `footpath`, `path`, `cycleway` or  `bridleway`, is determined as `nrOfLow` and the rest as `nrOfMajor`.

The following intersection types are assigned according to the number of different categories:

- If all `wayCount` ways belong to the categories `pedestrian`, `track`, `footpath`, `path`, `cycleway` or `bridleway` (`wayCount==nrOfLow`), the intersection node is classified as _low intersection_ (plotted as small rectangle).
- If at least three of these ways are major roads (`nrOfMajor> 2`), the intersection node is classified as _major intersection_ (plotted as large ring).
- If there are major roads, but less than three (`nrOfMajor< 3`), the intersection node is classified as _main intersection_ (plotted as large dot).

**Comment by vvoovv** [comment]
> For each node that is an intersection, the sum `wayCount` of all incoming and outgoing ways is determined.

Did you mean _way-segments_ or _ways_?

If a node is located at the middle of a way and there are no other ways that share the node, then `wayCount=1` and `waySegmentCount=2`

**Comment by polarkernel** [comment]
> Did you mean way-segments or ways?

I mean way-sections. I have never used the concept of ways that go through an intersection, because it can be ambiguous. If there are multiple way-sections at an intersection, it is not always clear, which ones belong to the same way.

**Comment by vvoovv** [comment]
> I'll think what to do with the filets in that context.

Created #83

**Comment by vvoovv** [comment]
> The question now is, where do the bundles (`Bundle` class) and the intersections (`Intersection` class) start? What about crosswalks? Do they belong to the bundles, and what about their intersections?

An intersection is created by finding the intersection points of the roadway borders. So a crosswalk is a part of the roadway.

But there is a notable exception: a diagonal crossing or a [pedestrian scramble ](https://en.wikipedia.org/wiki/Pedestrian_scramble):

![image](https://github.com/prochitecture/blosm/assets/613295/01f84c26-c680-473d-8238-5160af73dce3)


It's a part of a intersection and painted right on it. I suggest ignoring the minor ways (footways, crossways) when generating the intersections. Then the footways located within the intersection polygon can be filtered out.

I added the new OSM file _streets/tokyo_shibuya.osm_ which contains the famous[ Shibuya Crossing](https://en.wikipedia.org/wiki/Shibuya_Crossing).

**Comment by polarkernel** [comment]
> I added the new OSM file _streets/tokyo_shibuya.osm_ which contains the famous[ Shibuya Crossing](https://en.wikipedia.org/wiki/Shibuya_Crossing).

Very interesting crossing, but will be difficult to construct correctly:

<img src="https://github.com/prochitecture/blosm/assets/48180714/59fb771d-87da-4d16-b772-9ba11d4ddb27" width="400">

**Comment by vvoovv** [comment]
> Very interesting crossing, but will be difficult to construct correctly:

What will be the problems?

**Comment by vvoovv** [comment]
_streets/berlin_highways_without_intersections.osm_, the intersection at the middle.

![image](https://github.com/prochitecture/blosm/assets/613295/97eb65f5-c620-4898-8a2a-be94a4f9dcf6)

The darkest polygon includes the leftmost crossing node. However it is not the part of the intesection. Would it possible to fine tune the algorithm, so that the leftmost crossing node won't be included to the darkest intersection polygon?

**Comment by polarkernel** [comment]
> Would it possible to fine tune the algorithm, so that the leftmost crossing node won't be included to the darkest intersection polygon?

The current algorithm is still quite simple. It is based on a density-based spatial clustering algorithm ([DBSCAN](https://en.wikipedia.org/wiki/DBSCAN)), using the endpoints of all equidistant way-sections. It does not yet distinguish between the three intersection categories we have defined. An important parameter is the distance between cluster points, which is currently simply the Euclidean distance. This version has a tendency to connect clusters (or points) along strings of neighboring points, which seems to have happened in your example.

The algorithm has a lot of potential for improvement and control, which I will tweak as I get a better understanding of what the output should look like. For example, the distance can be weighted by the intersection types. There is also a hierarchical version of this algorithm. There is a lot of literature on this method of clustering.

**Comment by polarkernel** [comment]
> What will be the problems?

This is only a first intuitive impression.

**Comment by vvoovv** [comment]
I think the current results are enough to start generating some output in Blender. I suggest starting to generate the output using the new data structures. The algorithm to generate the intersections of bundles can be tweaked afterwards.

**Comment by polarkernel** [comment]
> I suggest starting to generate the output using the new data structures.

OK. First steps might be:
- Definition of the attributes and methods needed by the addon for rendering the `Section` class.
- Introduction of some kind of `getStyle` function, callable within `StreetGenerator`, which processes way-sections (`Section` class) by their PML style.
- I don't remember: Are still PML modifications required for `getStyle`? If so, where are they defined?
- Definition of a container for `Section` objects in the `WayManager`, which is filled by `StreetGenerator` as soon as they are fully processed and ready for rendering.

**Comment by vvoovv** [comment]
> * Definition of the attributes and methods needed by the addon for rendering the `Section` class.

https://github.com/prochitecture/blosm/issues/72#issuecomment-1905748582

> * Introduction of some kind of `getStyle` function, callable within `StreetGenerator`, which processes way-sections (`Section` class) by their PML style.

I will create an initial implementation of `getStyle` function and related PML.

> * I don't remember: Are still PML modifications required for `getStyle`? If so, where are they defined?

Modifications aren't needed right now. Generation of the streets is more important at the moment.

> * Definition of a container for `Section` objects in the `WayManager`, which is filled by `StreetGenerator` as soon as they are fully processed and ready for rendering.

What is the difference between this item and the first one?

**Comment by polarkernel** [comment]
> What is the difference between this item and the first one?

When a `Section` is processed by `getStyle`, it is returned to `StreetGenerator` for further processing. It can then be trimmed to fit to intersection connectors, be integrated into bundles, be invalidated, because it is too short, and the more. After that, it has to be transferred back to the addon for rendering, which requires some kind of container, I think, for this transport.

Of course, without the intersection areas, the streets can be rendered directly in the `getStyle` call for now.

**Comment by vvoovv** [comment]
The function `getStyle` will return a dictionary with the default values for some attributes (e.g. the number of lanes) and the values of the attributes that are not available in OSM at all (e.g. fillet radii),

What if the same instance of the class `Section` is used throughout the whole processing pipeline? It can have an attribute `valid` which is set to `False` if the section became invalidated.

**Comment by polarkernel** [comment]
> The function getStyle will return a dictionary with the default values for some attributes (e.g. the number of lanes) and the values of the attributes that are not available in OSM at all (e.g. fillet radii),

OK, I will wait for your proposition.

> What if the same instance of the class Section is used throughout the whole processing pipeline? It can have an attribute valid which is set to False if the section became invalidated.

Then the instance of this `Section` is the data in an edge of the instance `self.waymap` (`WayMap` class) in `StreetGenerator`. No problem, then, `WayMap` is the container I am talking about. If later the renderer needs to access a specific instance of `Section`, an adequate access function will have to be provided by `WayMap` and access to `self.waymap` has to be provided.

But we don't have to define that right now. We can work our way forward one step at a time.

**Comment by vvoovv** [comment]
Are the attributes `default`, `lane`, `doubleRoadsideWidth` from the dictionary `wayCategoryProps` at _way/way_properties.py_ used now in the code?

**Comment by polarkernel** [comment]
> Are the attributes default, lane, doubleRoadsideWidth from the dictionary wayCategoryProps at way/way_properties.py used now in the code?

`lane` is currently used in the function `estimateWayWidths()` as an estimate for the lane width, when no `width` is given in the section tags. The other two (`default` and `doubleRoadsideWidth`) are presently not used. The only additional attribute in the code is `nrLanes`, which is actually used in the function `lanePattern()`. Both functions are in the file _way/way_properties.py_.

**Comment by vvoovv** [comment]
> I will create an initial implementation of `getStyle` function and related PML.

It has been committed.

An additional parameter will be required in the command line:  `--assetsDir /path/to/folder/assets`.

Currently `--assetsDir` should point to the unpacked attached [file](https://github.com/prochitecture/blosm/files/14114931/assets.zip). It contains the PML file for streets: `assets/default/style/street/main.pml`.

I introduced the class `Street` in the module `way.item.street`. It should serve as a sort of wrapper for everything between two intersections. 

The `StreetGenerator` constructor now accepts two parameters: `styleStore` and `setStyle`. They are set as the attributes of the `StreetGenerator` instance in the constructor. An example code can be found between the lines 129 and 138 of the file _blosm/setup/mpl_blosm.py_.

A way should be proposed to access the first and the last items in a `Street`. In the example code the first item (`Section`) is accessed via the attribute `start`.

Once a `Street` will be populated with items, I will finish assigning the style for all street items in the method `Street.setStyle(..)` of the module `way.item.street`.

I renamed _way.items_ to _way.item_.

I added the parent class `Item` for all classes in _way.item_. However I think not all classes in _way.item_ require the parent class `Item`.

**Comment by vvoovv** [comment]
The default number of lanes and other default parameters are defined in the style block `street` for a given style. The attribute `totalNumLanes` and, where appropriate, `totalNumLanesOneway` are used for that.

**Comment by polarkernel** [comment]
> The default number of lanes and other default parameters are defined in the style block `street` for a given style. The attribute `totalNumLanes` and, where appropriate, `totalNumLanesOneway` are used for that.

During my research on lane patterns, I found that finding the number of lanes is a difficult process. In the current code, this is done by finding a lane pattern in the function `lanePattern()` ([line 171 in /way/way_properties.py](https://github.com/prochitecture/blosm/blob/5e8937c6391e9d72a37ea14db1027a6e046439ee/way/way_properties.py#L171)). The attributes      `isOneWay`, `totalLanes`, `forwardLanes`, `backwardLanes` and `bothLanes` are then derived from this pattern in the `Section` class. Additionally, the instances of the classes `SymLane` and `SideLane` will be constructed using this pattern. Now, I do not understand how to combine this pattern with the default parameters from the style block. What is `totalNumLanesOneway`?

There are two things we should not forget at this time:
- The meaning of the number of lanes is different for left-hand traffic and for right-hand traffic. In `lanePattern()`, this is taken into account by a function argument. But we will need a parameter (from the way-manager?), that gives this attribute.
- We have [defined some categories](https://github.com/prochitecture/blosm/issues/44#issuecomment-1265216790), that do not halve the way width, when they are tagged as one-way. This is implemented in the tuple `halveNrLanesIfOneWay` ([line 44 in /way/way_properties.py](https://github.com/prochitecture/blosm/blob/5e8937c6391e9d72a37ea14db1027a6e046439ee/way/way_properties.py#L44C1-L44C21)). I don't know if you want to keep these exceptions.

**Comment by polarkernel** [comment]
> A way should be proposed to access the first and the last items in a Street. In the example code the first item (Section) is accessed via the attribute start.

Something like this?

<img src="https://github.com/prochitecture/blosm/assets/48180714/83c5d4d3-530a-404b-ad76-cb784097e9fc" width="800">

Forward (start to end) should correspond to drawing direction of OSM operator, to keep correct lane directions.

**Comment by vvoovv** [comment]
> Now, I do not understand how to combine this pattern with the default parameters from the style block.

The values from the dictionary `wayCategoryProps` should be replaced by the values from a PML style block. 

> What is `totalNumLanesOneway`?

If `street.getStyleBlockAttr("totalNumLanesOneway")` returns `None`, then the total number of lanes is equal to street.getStyleBlockAttr("totalNumLanes") if it's a oneway section.

**Comment by vvoovv** [comment]
I created a temporary repository to host PML styles for streets: https://github.com/prochitecture/assets

**Comment by vvoovv** [comment]
> Something like this?

Exactly.

> Forward (start to end) should correspond to drawing direction of OSM operator, to keep correct lane directions.

There can be more complex cases.

A `Street` can also contain a `Bundle`. 

Also consider the following `Street`: `Section` - `SymLane` - `Section`. Those sections could have the opposite directions.

**Comment by vvoovv** [comment]
I renamed `totalNumLanes` to `totalLanes` and `totalNumLanesOneway` to `totalLanesOneway` in the PML file for brevity and consistence with the naming in the code.

**Comment by polarkernel** [comment]
> Once a Street will be populated with items, I will finish assigning the style for all street items in the method Street.setStyle(..) of the module way.item.street.

`Street` is now integrated into `StreetGenerator` for the current state (it contains only one instance of `Section`).  `getStyle()` is called for all such streets. The code is committed.

**Comment by polarkernel** [comment]
I created a version with debug output extracted to a file _debug.py_ and conditionally disabled, using `self.app.type`, as described [here](https://github.com/prochitecture/blosm/issues/72#issuecomment-1875309397). `StreetGenerator` now also runs in Blender if needed.

**Comment by vvoovv** [comment]
I removed the temporary code.

Why do some instances of the `Street` class have the `end` attribute equal to _None_?

**Comment by polarkernel** [comment]
> Why do some instances of the `Street` class have the `end` attribute equal to _None_?

Was a bug, fixed now.

**Comment by vvoovv** [comment]
How can I access instances of the `Street` class in the `WayManager` to render the streets in Blender?

**Comment by polarkernel** [comment]
> How can I access instances of the Street class in the WayManager to render the streets in Blender?

We didn't finally define this container after [this post](https://github.com/prochitecture/blosm/issues/80#issuecomment-1906421622). For now, I have defined the instance `waymap` of the class `WayMap` in the `WayManager` class. Currently, all edges in `waymap` are instances of `Street`, but later, they may also be instances of `Bundle`. You may iterate through all edges of the way-map using
```Python
for src, dst, key, street in manager.waymap.iterSections():
    ...
```
`src` and `dst` are the ends (`mathutils.Vector`) of the street. `key` is usually 0, but is increased when multiple streets have the same end points. `street` is an instance of the class `Street`. When once `Street` and `Bundle` both will be mixed in `waymap`, you may select all streets by
```Python
for src, dst, key, street in manager.waymap.iterSections(Street):
    ...
```
If you know the endpoints `src` and `dst` of the street, you may access all edges in `waymap` using
```Python
for edge in manager.waymap.getSections(src, dst):
    ...
```
or, when you know the key of the edge
```Python
edge = manager.waymap.getSectionObject(src, dst, key)
```
But the last two methods of `WayMap` do not distinguish the type of the edge (may be `Street` or `Bundle`).

**Comment by vvoovv** [comment]
But a `Bundle` itself is a part of a Street as described in the original [message](https://github.com/prochitecture/blosm/issues/72#issuecomment-1846141098).

**Comment by polarkernel** [comment]
> But a `Bundle` itself is a part of a Street as described in the original [message](https://github.com/prochitecture/blosm/issues/72#issuecomment-1846141098.).

You are right, I missed that. For now, it is not a problem, because there are only instances of `Street` in `waymap` that contain one single `Section`.

Looking back at these definitions, I still have trouble understanding (and accepting) them. In my opinion, these structures do not correctly reflect the topology of the road network. They make the solution much more complicated than needed. Let me take the example of a `Street` implementation above:

<img src="https://github.com/prochitecture/blosm/assets/48180714/83c5d4d3-530a-404b-ad76-cb784097e9fc" width="800">

The `Street` object is a collection of linear objects (`Section`, `SymLane`) that are connected as a doubly linked list. Each of these objects has two ends connected to the neighboring objects.. The `Street` object itself also has two ends, connected to intersections, which are the natural limits of a `Street`. The `Intersection` objects are different. They are not linear, they have multiple connections, and they have an area.

The example from the [message](https://github.com/prochitecture/blosm/issues/72#issuecomment-1846141098.) violates this characteristic:

<img src="https://github.com/prochitecture/blosm/assets/613295/5b206ec8-3c67-4dd9-981b-346b0930a331" width="650">

The `SplitMerge` (SM) object has three connectors and is more like an intersection than a section. Possibly even an area is required to represent it. The `Bundle` has two ends at the right (and could have multiple ends in general case). Topologically, they connect to different connectors of the intersection. It is the `Bundle` with its details, that connects to the intersection, and not the `Street`. Topologically, this construct is a graph, with a node SM in addition to the intersections.

I have even more trouble with the example, that we already discussed [earlier](https://github.com/prochitecture/blosm/issues/72#issuecomment-1847628177):

<img src="https://github.com/prochitecture/blosm/assets/48180714/95a60e12-03b2-4ba2-afe4-fd23e8826b7a" width="500">

The introduction of a `PartialIntersection` does not help. To represent this construction, `Street` needs an additional attribute besides `start` and `end`. It is a subgraph of the street network and not a linear structure.

These considerations led me to choose a graph as the basic structure in `StreetGenerator`. For now, we can keep the described interface to access the `Street` objects. However, maybe I may have to adapt it later.

**Comment by vvoovv** [comment]
There are issues to solve if a graph is accepted as the basic structure in `StreetGenerator`.

Is `Crosswalk` a node or an edge of the graph?

Is `PtStop` a node or an edge of the graph?

What is the relation between a `Bundle` and the nodes and the edges of the graph? Grouping into Bundles is also needed to generate separators (grass, etc) between the roadways that form a `Bundle`.

**Comment by polarkernel** [comment]
> Is `Crosswalk` a node or an edge of the graph?
Is `PtStop` a node or an edge of the graph?

What are they in the doubly linked list of a `Street`?

> What is the relation between a `Bundle` and the nodes and the edges of the graph?

In the old solution, `Bundles` have been `LongClusterWays`, linear structures that ended at a (complicated) `Intersection`. The issue of `PartialIntersections` had been solved by splitting a `LongClusterWay` into `SubClusters` using a `SplitDescriptor`. So something like `SubCluster` could have been an edge of the graph and `SplitDescriptor` and `Intersection` been nodes.

However, I would like to change the procedure at this point. I have been trying to find solutions for this for years now, but have always failed. Either the problem is too complicated for me or I'm too stupid. I now feel tired and don't want any more to carry on working like this.

You seem to have an idea of how you would solve it. Also, you are looking at the problem from a different viewpoint than I am, from a rendering perspective. This can lead to new ideas and solutions. I suggest you tell me how you would like to have what structures, step by step, and I will adapt. If necessary, I will replace the graph (`waymap`) with something else (currently it could be just a dictionary).

Can you cope with that?

**Comment by vvoovv** [comment]
I'd like to have the structure presented in the original [message](https://github.com/prochitecture/blosm/issues/72#issuecomment-1846141098). This structure would allow creating the separator areas by connecting the internal borders of the items and forming the polygons made of those internal border chunks.

Green numbers (1) and (2) on the image below represent the separator areas. 

![image](https://github.com/prochitecture/blosm/assets/613295/d68ba123-dfa3-474d-ab14-c9bd370022b2)

Those internal border chunks are shown with the green color of two shades on the image below:

![image](https://github.com/prochitecture/blosm/assets/613295/4205ed26-5b36-44ed-a0f3-d68b1b411539)


Finally, I think a separate `PartialIntersection` class is not needed. The single class `Intersection` would be enough.

**Comment by polarkernel** [comment]
> I'd like to have the structure presented in the original [message](https://github.com/prochitecture/blosm/issues/72#issuecomment-1846141098).

OK, let me describe that in my own words so far, so that you can check, whether I understand it. 

What I see are base objects (`Section`, `PtStop`, `Intersection`, `SplitMerge`, `Crosswalk`). 
- A `Section` is an elongated object, that has a centerline and two connectors, one at each end. Its attributes are already well-defined in the corresponding class of the current code.
- A `PtStop` could be something similar to a `Section`, but with additional attributes. I suggest leaving its definition open until we start implementing it. See also #76.
- An `Intersection` is a polygonal object, that has two or more connectors.
- A `SplitMerge` looks like an `Intersection` to me. It has three connectors. Maybe it has a direction. Again, I suggest leaving its definition open until we start implementing it.
- A `Crosswalk` will require a special structure. Some ideas are already discussed in #73.

We need to discuss how these objects are connected and how they are passed from `StreetGenerator` to `WayManager`.

> Green numbers (1) and (2) on the image below represent the separator areas.

These separators will only be valid between parallel way-sections, not when the road runs around a city block. The concept of a `Bundle` will be required for this, but like `Street`, this is no longer a base object.

**Comment by vvoovv** [comment]
> OK, let me describe that in my own words so far, so that you can check, whether I understand it.

Yes, everything is correct with some minor notes.

> What I see are base objects (`Section`, `PtStop`, `Intersection`, `SplitMerge`, `Crosswalk`).

I think the objects `TransitionSymLane` and `TransitionSideLane` are also the base ones.

`TransitionSymLane` is like an `Intersection`, but it has only two connectors. A `Street` can't end at a `TransitionSymLane`.

Also a `Street` can't end at a `SplitMerge`.

`TransitionSideLane` is like an `Intersection` with zero area, but it has only two connectors. A `Street` can't end at a `TransitionSideLane`


> * An `Intersection` is a polygonal object, that has two or more connectors.

Doesn't it have _three_ or more connectors?

**Comment by polarkernel** [comment]
> Yes, everything is correct with some minor notes.

Thanks for the additions, they complete the set of base objects.

> * An Intersection is a polygonal object, that has two or more connectors.

> Doesn't it have three or more connectors?

Maybe, there is a special case at corners. I will post it as a new issue today.

**Comment by polarkernel** [comment]
> Maybe, there is a special case at corners. I will post it as a new issue today.

Done in #84.

**Comment by vvoovv** [comment]
Another reason of having Bundles is the generation of sidewalks.

Without a Bundle the sidewalks are generated along both sides of a Section or another item. 

If a Bundle is present, the sidewalks are generated along both sides of the Bundle (or the top Bundle if there is a hierarchy of Bundles), rather than the sides of each member of the Bundle.

**Comment by vvoovv** [comment]
> ```python
> for src, dst, key, street in manager.waymap.iterSections():
>     ...
> ```

I started a rewrite of the street renderer. For now I managed to render instances of the class `Section` using the above method.

![image](https://github.com/prochitecture/blosm/assets/613295/f839d398-0fab-4d9e-a9a3-2c85220c4343)

Let's discuss what data can be delivered to the renderer in the next step of the development.

**Comment by polarkernel** [comment]
> Let's discuss what data can be delivered to the renderer in the next step of the development.

I see three different approaches:

- Implement first `items` within a `Street`: `SideLane` and `SymLane`. This would be interesting, because we could discuss and find out how to extend the doubly linked list in a `Street`.
- Implement `PtStop`. Requires detecting them, finding a way to hold them in a `Street`, and finally rendering them. This would be a very demanding task, but is a fairly open point of the structures.
- After the edges, implement the nodes in the `waymap`: `Intersections`. Could be based on the old implementation, using their connectors and the trimming of the way-sections. Nothing really new, but would "beautify" the rendered result. No clustered intersections for now.

**Comment by vvoovv** [comment]
I'd prefer to start from the intersections.

What are the obstacles now to get clustered intersections?

**Comment by polarkernel** [comment]
> What are the obstacles now to get clustered intersections?

Clustered intersections require not only the definition and construction of clusters, but also those of bundles. This will be a huge and demanding development step. See also the discussion starting [here](https://github.com/prochitecture/blosm/issues/80#issuecomment-1880045995). To get illustrations, you can run the code in script mode, using `--highways --generateStreets`.

**Comment by polarkernel** [comment]
> I'd prefer to start from the intersections.

In the old implementation, the neighbors of intersections were way-sections. Their boundaries were intersected, as shown below, and these intersection points (red dots) were connected to a polygon (red) that was completed so that the connectors were perpendicular. The way-sections were then trimmed along their centerlines so that the connectors fit the intersection. 

<img src="https://github.com/prochitecture/blosm/assets/48180714/dccbe384-e46d-4d4a-944e-115fd50e907e" width="300">

With the new structures, the neighbors of an instance of `Intersection` are instances of `Street`, connected to their `start` or `end`. Also, an intersection cluster at the tails of `Bundles` will be an instance of `Intersection`, and similar rules apply. Therefore, instances of `Street` and of `Bundle` <ins>must end in an instance of </ins> `Section`, otherwise, this algorithm can't be applied. We will need a solution for `Crosswalks`.

**Comment by vvoovv** [comment]
> Therefore, instances of `Street` and of `Bundle` must end in an instance of `Section`, otherwise, this algorithm can't be applied. We will need a solution for `Crosswalks`.

What if `Crosswalks`, `PtStops` and other similar items are inserted into a `Street` after the algorithm finished its job? In this manner the algorithm will deal with `Sections` only, as the neighbors of `Intersections`.

**Comment by polarkernel** [comment]
> What if `Crosswalks`, `PtStops` and other similar items are inserted into a `Street` after the algorithm finished its job? In this manner the algorithm will deal with `Sections` only, as the neighbors of `Intersections`.

This is a viable option. The only condition is, that the perpendicular connector line of the intersection still fits to whatever has been inserted. 

Another point is the connection between `Intersection` and `Street`. I think, my first idea will not work:

<img src="https://github.com/prochitecture/blosm/assets/48180714/83c5d4d3-530a-404b-ad76-cb784097e9fc" width="800">

The `Intersection`'s connector should reference the instance of `Street` and, additionally, an information on which end of this instance. Instead of `pred` and `succ` of the `Sections`, new attributes `pred` and `succ` of the instance of `Street` should reference the instance of the `Intersection`, and additionally, their connector. When `Crosswalks`,  `PtStops` and other similar items are inserted into a `Street`, only the attributes `start` and `end` have to be adapted, while the instance of `Intersection` remains unchanged.

**Comment by vvoovv** [comment]
> I think, my first idea will not work

Then `s1.pred = s2.succ = None` to indicate that `s1` and `s2` are the first and the last items in the `Street`?

**Comment by polarkernel** [comment]
> Then s1.pred = s2.succ = None to indicate that s1 and s2 are the first and the last items in the Street?

Yes. I also propose to rename `start` and `end` to `head` and `tail`, as it is usually done for doubly linked lists. `succ` and `pred` of `Street` would then reference the `Intersection` before and after the `Street`.

Next issue: How can one find the connector to an `Intersection`, when the street is given, and how can one reference the street that belongs to a connector, when the intersection is given? I would like to keep the idea to use direct references to objects and not to use any index or similar. Let me illustrate that using the image below.

<img src="https://github.com/prochitecture/blosm/assets/48180714/a6dba9a6-44cd-4664-a784-54f9ca605fab" width="800">

Given the intersection `is1`, one of its connectors, for instance `c2`, could contain a tuple `c2=(st1, i)`, where `st1` is the reference to the street `st1` and `i` the index of a point on the connector line of the intersection's polygon, as we had it already before. Similarly, the sign of `i` could indicate the start or the end of `st1`. 

The inverse direction is more complicated. The reference `pred` of `st1` lets us find the instance `is1` of the intersection. But how can we find the corresponding connector `c2`? One could implement `connectors` as a Python list and add the index 2 to `pred` of `st1`. But the net may be modified, streets be removed or be combined to bundles, so that this index must be considered as mutable.

I had an idea, that may look strange in the first moment. I am also not really sure that it would work under all circumstances. It is possible to use an object as key of a dictionary in Python, given it has a `__hash__()` and a `__eq__()` method. These could be implemented for our objects, using for instance `hash(location)` for intersections or `hash( (src,dst,key) )` for elongated objects. Using this feature, `connectors` would be implemented as a dictionary, using `connectors[st1]=i` to reference `st1`, while `i=st1.pred.connectors[st1]` would reference the connector to `st1`.

Do you have any ideas? Or do you maybe have less or additional requirements, seen from the viewpoint of the renderer?

**Comment by vvoovv** [comment]
> Yes. I also propose to rename `start` and `end` to `head` and `tail`, as it is usually done for doubly linked lists.

Ok.

> Next issue: How can one find the connector to an `Intersection`, when the street is given, and how can one reference the street that belongs to a connector, when the intersection is given?

Another possible solution is to use a separate class for the connectors and to get rid of the connector indices all together. 

```python
class IntConnector:

    def __init__(self, intersection):

        # the intersection, to which the connector belongs to
        self.intersection = intersection

        # the item, to which the connector is connected to
        self.item = None

        # the preceding connector in the intersection (for, example, when considered counterclockwise)
        self.pred = None

        # the succeeding connector in the intersection (for, example, when considered counterclockwise)
        self.succ = None
```

`street.tail` and `street.head` are then set to a instance of the class `IntConnector`.

**Comment by polarkernel** [comment]
> Another possible solution is to use a separate class for the connectors and to get rid of the connector indices all together.

This is a great idea, I like it. A circular doubly-linked list of instances of `IntConnector`. The class `Intersection` would need an anchor to hold a start of this list, for example `connectors`. Similar to what we already did in the old implementation, within `IntConnector`, the attribute `self.index` is the index of a vertex in the intersection polygon, so that the item connects between `self.index` and `self.index+1`. Either the sign of this index or the additional attribute `self.fwd`, which is `True`, if the start of the street is connected, and `False` otherwise, define the direction of the street. I would prefer the latter. The order of the polygon in `Intersection` is counter-clockwise, so it makes sense to order this circular list in the same direction. The class definition becomes then
```python
class IntConnector:

    def __init__(self, intersection):

        # the intersection, to which the connector belongs to
        self.intersection = intersection

        # the first index of the connector in the area polygon of Intersection
        self.index = None

        # the item, to which the connector is connected to
        self.item = None

        # the direction of the item, to which the connector is connected to
        self.fwd = None

        # the preceding connector in the intersection (in clockwise direction)
        self.pred = None

        # the succeeding connector in the intersection (in counter-clockwise direction)
        self.succ = None
```
OK for you?

> `street.tail` and `street.head` are then set to a instance of the class `IntConnector`.

You mean `street.pred` and `street.succ` in my image.

Once all the items are set up by the `StreetGenerator`, do you already have an idea of how you want to access them? It is not possible to just provide a starting item and then to follow the links, because the network may contain islands.

**Comment by vvoovv** [comment]
> The class `Intersection` would need an anchor to hold a start of this list, for example `connectors`.

Is a list of connectors? I suggest having an attribute with the name `startConnector` or `firstConnector` that holds a reference to an initial connector.

> Either the sign of this index or the additional attribute `self.fwd`, which is `True`, if the start of the street is connected, and `False` otherwise, define the direction of the street. I would prefer the latter.

`self.index` can be also equal to zero, so it can't be used to define the direction of a street. I think the term "incoming" is used throughout the code. I suggest naming it `self.incoming`.

> OK for you?

Yes, with the notes above and below.

> You mean `street.pred` and `street.succ` in my image.

Yes.

> Once all the items are set up by the `StreetGenerator`, do you already have an idea of how you want to access them?

`for src, dst, key, street in manager.waymap.iterSections()` seems to be ok for accessing the streets. There is a special case of a circular street. In that case `street.tail` is equal to `street.head`. The attributes `street.pred` and `street.succ` aren't used.

> It is not possible to just provide a starting item and then to follow the links, because the network may contain islands.

Doesn't `manager.waymap.iterSections()` return all `Streets` if there are islands?

**Comment by polarkernel** [comment]
> Is a list of connectors? I suggest having an attribute with the name startConnector or firstConnector that holds a reference to an initial connector.

It is a circular doubly-linked list. The attribute `startConnector` of `IntConnector` holds a reference to the first connector in the list:

<img src="https://github.com/prochitecture/blosm/assets/48180714/3f0b7f74-3f09-4a24-9346-aedcbc8d3cf8" width="700">

> `self.index` can be also equal to zero, so it can't be used to define the direction of a street. I think the term "incoming" is used throughout the code. I suggest naming it `self.incoming`.

You are right. To be consistent with the existing code, I named it `self.leaving`. If this is `True`, this means that the start of the street is at the intersection. The sequence of the items in `Street` start at `head`.

> `for src, dst, key, street in manager.waymap.iterSections()` seems to be ok for accessing the streets. There is a special case of a circular street. In that case `street.tail` is equal to `street.head`. 

Yes.

> The attributes `street.pred` and `street.succ` aren't used.

Yes, if it is a circular street with no intersection. For a loop, which is also kind of circular, these must reference their intersections

> Doesn't `manager.waymap.iterSections()` return all `Streets` if there are islands?

Yes, it returns all `Streets` in the network.

**Comment by vvoovv** [comment]
I think everything is agreed now. Looking forward to the intersections in the street renderer.

**Comment by polarkernel** [comment]
> I think everything is agreed now. Looking forward to the intersections in the street renderer.

All this is already implemented in the current version. Most of the intersections already look as desired, when executed in script mode:

![1](https://github.com/prochitecture/blosm/assets/48180714/6422ec8b-f48b-486b-a9f8-eaa723ec1493)

but there are overlaps, because we do not yet have any clustering, which can produce some "salad" like this

![2](https://github.com/prochitecture/blosm/assets/48180714/f4a7e7a2-e6c0-455a-9e4e-37e5db136ee1)

There also seem to be some minor bugs in the construction of the intersections. However, the structures should already be correct (though not yet thoroughly tested), so you might want to do some initial checks.

**Comment by vvoovv** [comment]
Would it be possible to enable trimming of centerlines due to the intersections?

**Comment by polarkernel** [comment]
> Would it be possible to enable trimming of centerlines due to the intersections?

Committed. It is a very temporary solution, implemented only for your checks. It will be solved elsewhere, once I finalize the code. 

Note: `Section` has become an attribute `valid`. Use only valid `Sections`.

**Comment by vvoovv** [comment]
> Note: `Section` has become an attribute `valid`. Use only valid `Sections`.

If a `Section` in a `Street` is invalid, does it mean that the whole `Street` becomes invalid?

**Comment by vvoovv** [comment]
Also, I'd like to have all `Intersections` in a Python list (e.g. `manager.intersections`). I think it's much easier to create it in the street generator, than trace the intersection in the renderer down, since the `Intersections` are created consequently in the street generator.

**Comment by polarkernel** [comment]
> If a `Section` in a `Street` is invalid, does it mean that the whole `Street` becomes invalid?

For now, yes. This is only a temporary solution, because I cannot handle all problems with short sections at the moment. Later, invalid sections will be removed, and the links will be adjusted somehow.

> Also, I'd like to have all `Intersections` in a Python list (e.g. `manager.intersections`).

I assumed that when I asked for "how you want to access them". I will provide that as soon as possible. Currently, only intersections with three or more leaving sections are processed. Later, the others will become `SymLanes`, `SideLanes`, `Corners`(?), ..., they are not yet included in this list.

**Comment by polarkernel** [comment]
>  I will provide that as soon as possible.

Committed.

**Comment by vvoovv** [comment]
I'll take some time to develop the code.

I'd suggest including `SymLanes` and `SideLanes` in the next step. I suggest not including them to `manager.intersections`, since there is a 1-to-1 mapping between a transition and its `Street`.

**Comment by polarkernel** [comment]
> I'd suggest including `SymLanes` and `SideLanes` in the next step. I suggest not including them to `manager.intersections`, since there is a 1-to-1 mapping between a transition and its `Street`.

OK, I intend to implement them as a linked instance between two `Sections` in a `Street`, as shown in the image[ here](https://github.com/prochitecture/blosm/issues/80#issuecomment-1943370866).

**Comment by vvoovv** [comment]
> OK, I intend to implement them as a linked instance between two `Sections` in a `Street`, as shown in the image[ here](https://github.com/prochitecture/blosm/issues/80#issuecomment-1943370866).

I meant exactly that.

**Comment by polarkernel** [comment]
> I'd suggest including `SymLanes` and `SideLanes` in the next step. I suggest not including them to `manager.intersections`, since there is a 1-to-1 mapping between a transition and its `Street`.

A first version has been committed. I think the new classes  `SymLane` and `SideLane` are almost self-explanatory with my comments. They are embedded in their `Street` in a double-linked list, as shown [here](https://github.com/prochitecture/blosm/issues/80#issuecomment-1943370866).

For `SideLane`, the incoming and outgoing sections are not trimmed at the transition vertex. The transition must be created in the addon. The widths of the incoming and outgoing sections are still corrected according to [this post](https://github.com/prochitecture/blosm/issues/57#issuecomment-1544025403).

A `SymLane` contains an attribute `area`, which is the polygon of its transition area. The incoming and outgoing sections are trimmed at the transition vertex to fit the transition area. I didn't add connectors as we did for `Intersections`. The incoming (smaller) section always connects to the vertex `area[0]` and the outgoing section to `area[2]` (as the first vertex in counter-clockwise direction).

In the current version, sections are not yet split by `Corners`, as discussed [here](https://github.com/prochitecture/blosm/issues/84#issue-2120442895), because we do not yet use parallel sections.

**Comment by polarkernel** [comment]
> I didn't add connectors as we did for Intersections.

Sorry, I forgot to remove some code that uses connectors in `SymLane`. A fixed version without this code is committed.

**Comment by polarkernel** [comment]
I committed a version with code for experimental intersection clustering. This is only for backup reasons, the code itself is not called.
